torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu

```diff
@@ -0,0 +1,173 @@
+#ifndef THC_GENERIC_FILE
+#define THC_GENERIC_FILE "generic/VolumetricAdaptiveAveragePooling.cu"
+#else
+
+#include "../common.h"
+
+// 5d tensor B x D x T x H x W
+
+void THNN_(VolumetricAdaptiveAveragePooling_updateOutput)(
+           THCState *state,
+           THCTensor *input,
+           THCTensor *output,
+           int osizeT,
+           int osizeW,
+           int osizeH)
+{
+  THCUNN_assertSameGPU(state, 2, input, output);
+
+  THCUNN_argCheck(state, input->nDimension == 4 || input->nDimension == 5, 2, input,
+                  "4D or 5D (batch mode) tensor expected for input, but got: %s");
+
+
+  real *output_data;
+  real *input_data;
+
+  int64_t sizeD, isizeT, isizeH, isizeW;
+  int64_t istrideD, istrideT, istrideH, istrideW;
+  int64_t totalZ;
+
+  if (input->nDimension == 4) {
+    sizeD = input->size[0];
+    isizeT = input->size[1];
+    isizeH = input->size[2];
+    isizeW = input->size[3];
+
+    istrideD = input->stride[0];
+    istrideT = input->stride[1];
+    istrideH = input->stride[2];
+    istrideW = input->stride[3];
+
+    THCTensor_(resize4d)(state, output, sizeD, osizeT, osizeH, osizeW);
+
+    totalZ = sizeD * osizeT;
+  } else {
+    input = THCTensor_(newContiguous)(state, input);
+
+    int64_t sizeB = input->size[0];
+    sizeD = input->size[1];
+    isizeT = input->size[2];
+    isizeH = input->size[3];
+    isizeW = input->size[4];
+
+    istrideD = input->stride[1];
+    istrideT = input->stride[2];
+    istrideH = input->stride[3];
+    istrideW = input->stride[4];
+
+    THCTensor_(resize5d)(state, output, sizeB, sizeD, osizeT, osizeH, osizeW);
+
+    totalZ = sizeB * sizeD * osizeT;
+  }
+
+  input_data = THCTensor_(data)(state, input);
+  output_data = THCTensor_(data)(state, output);
+
+  int64_t offsetZ = 0;
+  dim3 threads(32, 8);
+  // each H*W plane is processed by blocksH thread blocks
+  int blocksH = max((int)(16L / totalZ), 1);
+  while (totalZ > 0) {
+    dim3 blocks(totalZ > 65535 ? 65535 : totalZ, blocksH);
```

Comment:
why did you choose to break things up this way (blocks,threads) (where the blocks are 2d), instead of the existing Volumetric pooling (VolumetricAveragePooling, VolumetricDilatedMaxPooling) where it's (grid/blocks) and grid is 3d?

Reply:
`(blocks,threads)` naming is following the SpatialAdaptiveAveragePooling.
That blocks is 2d is also following SpatialAdaptiveAveragePooling, where each thread does multiple pixels rather than 1 (`blocksH` blocks together process a H*W plane). In VolumetricDilatedMaxPooling, H*W/(32*8) blocks process a plane. 

I'm not sure if one way is better than the other. But I thought following spatial version should be a reasonable thing to do. :)